File: Python/128__Longest_Consecutive_Sequence.py
## This is a straight forward solition, that we sort the nums first,
## then check one by one. Runtime is O(N * lgN)


class Solution(object):
    def longestConsecutive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0

        nums.sort()

        longest_streak = 1
        current_streak = 1

        for i in range(1, len(nums)):
            if nums[i] != nums[i-1]:
                if nums[i] == nums[i-1] + 1:
                    current_streak += 1
                else:
                    longest_streak = max(longest_streak, current_streak)
                    current_streak = 1

        return max(longest_streak, current_streak)


## Another solution is brutal force, but use set() that have O(1) look
## up cost. The runtime is optimized to O(N)

class Solution:
    def longestConsecutive(self, nums):
        longest_streak = 0
        num_set = set(nums)

        for num in num_set:
            if num - 1 not in num_set:
                current_num = num
                current_streak = 1

                while current_num + 1 in num_set:
                    current_num += 1
                    current_streak += 1

                longest_streak = max(longest_streak, current_streak)

        return longest_streak

File: Java/128__Longest_Consecutive_Sequence.java
class Solution {
    public int longestConsecutive(int[] nums) {
        Set<Integer> data = new HashSet<>();
        for (int num : nums) {
            data.add(num);
        }
        HashMap<Integer, Integer> map = new HashMap<>();
        int max_len = 0;
        for (Integer val : data) {
            boolean t1 = map.containsKey(val - 1);
            boolean t2 = map.containsKey(val + 1);
            if (t1 && t2) {
                int start = map.get(val - 1);
                int end = map.get(val + 1);
                map.remove(start);
                map.remove(val - 1);
                map.remove(end);
                map.remove(val + 1);
                map.put(start, end);
                map.put(end, start);
            } else if (t1) {
                int start = map.get(val - 1);
                int end = val;
                map.remove(start);
                map.remove(val - 1);
                map.put(start, end);
                map.put(end, start);
            } else if (t2) {
                int start = val;
                int end = map.get(val + 1);
                map.remove(end);
                map.remove(val + 1);
                map.put(start, end);
                map.put(end, start);
            } else {
                map.put(val, val);
                map.put(val, val);
            }
        }

        for (Integer key : map.keySet()) {
            if (max_len < Math.abs(key - map.get(key)) + 1) {
                max_len = Math.abs(key - map.get(key)) + 1;
            }
        }
        return max_len;
    }
}

// Easier Solution
class Solution {
    public int longestConsecutive(int[] nums) {
        Set<Integer> data = new HashSet<Integer>();
        for (int num : nums) {
            data.add(num);
        }

        int max_len = 0;

        for (int num : num_set) {
            if (!data.contains(num - 1)) {
                int currentNum = num;
                int currentStreak = 1;
                while (data.contains(currentNum + 1)) {
                    currentNum += 1;
                    currentStreak += 1;
                }
                max_len = Math.max(max_len, currentStreak);
            }
        }

        return max_len;
    }
}
